[IPAM] NetConfig, Reservation and IPSet CRD

stuggi opened the pull request:
Adds CRDs for IPAM support
* NetConfig - describes the overall networks
* IPSet - requestor requests reservations for a list of networks
  and subnet. Also a fixed IP can be requested which does not have
  to be part of the AllocationRanges of a subnet.
* Reservation - "internal" store of the reservations for an IPSet.
  Allows to backup/restore reservations.

Follow up TODOs:
* add validation/defaulting to the webhooks for netcfg, ipset and reservation
  - validate input
    - allocationRanges match the cidr
    - excludeAddresses match the cidr
  - validate which parameters must not change on an update
  - validate ipset against netcfg requested networks/subnets are valid
  - ipset flagged defaultRoute net has a gw
  - ipset fixedIP requested is available
  - reservation ips in a subnet should not change on an update
  
Jira: [OSP-19996](https://issues.redhat.com//browse/OSP-19996)

stuggi commented:
@rabi @hjensas I made some bigger changes to what I shared with you earlier this week.
* the func which does the ip assignment is almost rewritten and I think its now cleaner and small (not tested yet with ipv6)
* the netcfg controller is obsolet. I think we can remove it and just have webhook for validation/defaulting (need to remove it)
* the reservation webhook is also not really required, only if we want to have the current status I added there to show the reservations in the cli
```
$ oc get reservations                                                                                                                                                       
NAME               READY   MESSAGE          RESERVATION                                                                                                                                                             
edpm-compute-001   True    Setup complete   ["CtlPlane: 192.168.122.100","InternalApi: 172.17.0.11","Storage: 172.18.0.10","Tenant: 172.20.0.10"]                                                                   
edpm-compute-002   True    Setup complete   ["CtlPlane: 192.168.122.101","InternalApi: 172.17.0.13","Storage: 172.18.0.11","Tenant: 172.20.0.11"]                                                                   
edpm-compute-003   True    Setup complete   ["CtlPlane: 192.168.122.102","InternalApi: 172.17.0.14","Storage: 172.18.0.12","Tenant: 172.20.0.12"]                                                                   
edpm-compute-004   True    Setup complete   ["CtlPlane: 192.168.122.103","InternalApi: 172.17.0.15","Storage: 172.18.0.13","Tenant: 172.20.0.13"]                                                                   
edpm-compute-005   True    Setup complete   ["CtlPlane: 192.168.122.104","InternalApi: 172.17.0.16","Storage: 172.18.0.14","Tenant: 172.20.0.14"]
```
could also just have a status in the ipset and also remove the reservation controller and just have webhook if needed. I have tested with creating 100 ipsets via the script in `test/scripts/test-ipset.sh`. This approach was a lot faster in mass create of the reservations then the previous design with the netcfg controller to watch for ipsets and handle the reservation create.

stuggi commented:
ignore the dns related commits ... was not good that I based it on them for now. the latest commit is relevant https://github.com/openstack-k8s-operators/infra-operator/pull/40/commits/04f6a4093e5735f36488516bea018d25bfdb3763

hjensas commented on controllers/network/ipset_controller.go:
```diff
@@ -0,0 +1,394 @@
+	// reservations map map[string]Reservation with reservation name as index
+	resMapNameIdx := make(map[string]networkv1.Reservation)
+	// map with all ip -> hostname references
+	resMapIPIdx := make(map[string]string)
```
While convenient, this map looks like a potential hot-spot at scale?
If I undestrand it correctly it will iterate over all IPs already reserved in all networks for each individual new allocation?

I think we can use the `slices.IndexFunc` in `pkg/ipam/funcs.go`'s `fixedIPExists()` and `iterateForAssignment()`. If we can operate directly against `ReservationList`  there instead I believe we would avoid iterating trough everything every time?

Like this:
```
	f := func(c Reservation) bool {
		for _, v := range c.Spec.Reservations {
			if v.Address == nextip {
				return true
			}
		}
		return false
	}
	idx := slices.IndexFunc(reservations.Items, f)
	if  (idx >= 0 && reservations.Items[idx].Spec.Hostname != a.Hostname) {
		fmt.Println(fmt.Sprintf("IP: %s already reseverd for host: %s", nextip, a.Hostname))
	}

```

Or, if we can add NetName name to `AssignIPDetails` we could do:
```
	f := func(c Reservation) bool {
		return c.Spec.Reservations[a.NetName].Address == nextip
	}
	idx := slices.IndexFunc(reservations.Items, f)
	if (idx >= 0 && reservations.Items[idx].Spec.Hostname != a.Hostname) {
		fmt.Println(fmt.Sprintf("IP: %s already reseverd for host: %s", nextip, a.Hostname))
	}
```

hjensas reviewed (commented):
Overall looks good. Thanks for switching to use net/ipnet and the next() function etc. I think using standard library makes sense and definitely smaller and easier to read.

I have made some comments in line.

stuggi commented on apis/network/v1beta1/reservation_types.go:
```diff
@@ -0,0 +1,94 @@
+)
+
+// IPAddress -
+type IPAddress struct {
```
yes, the initial plane was to reduce lookup churn. Instead of adding it here, I'll move it to the status of the ipset, so that when the reservation is done all information is reflected in the IPSet to the requestor.

stuggi commented on config/samples/network_v1beta1_ipset.yaml:
```diff
@@ -0,0 +1,15 @@
+  name: edpm-compute-0
+spec:
+  hostname: edpm-compute-0
+  networks:
```
yes, that sounds good

stuggi commented on pkg/ipam/funcs.go:
```diff
@@ -0,0 +1,124 @@
+			}
+
+			if hostname, exist := a.Reservelist[nextip.String()]; exist && hostname != a.Hostname {
+				helper.GetLogger().Info(fmt.Sprintf("%s already reserved for %s", nextip.String(), hostname))
```
yes, that was more for me during dev of that

stuggi commented on pkg/ipam/funcs.go:
```diff
@@ -0,0 +1,124 @@
+}
+
+// GetCidrParts - returns addr and cidr suffix
+func GetCidrParts(cidr string) (string, int, error) {
```
right, its a left over

stuggi commented on controllers/network/ipset_controller.go:
```diff
@@ -0,0 +1,394 @@
+	}
+
+	// reservations map map[string]Reservation with reservation name as index
+	resMapNameIdx := make(map[string]networkv1.Reservation)
```
yes, can be dropped

stuggi commented:
@hjensas I think I addressed all/most of your comments. Thanks for those! A lot of the complexity in reading is now removed. I think its small and clean. For this initial version I will do tomorrow
- add more envtest test cases
- enhance the ipset controller when there are no more IPs in an Allocation range. the condition still reports complete with no reservation.
- add all relevant information for the consumer to the ipset status, so that all network details could fetched from there instead of query the netconfig for it, you'd agree on that?

As follow up PRs/tasks
- add validation/defaulting to the webhooks
    - netcfg
        -  cidr - ok
        -  allocationRanges match the cidr
        - excludeAddresses match allocationRange
        - which information should not be possible to be changed after initial create to not mess up reservations?
    - ipset 
        - networks/subnet against netcfg
        - flagged defaultRoute net has a gw
        - fixedIP requested is available
        - which information should not be possible to be changed after initial create?
    - reservation 
        - ips in a subnet should not change on an update
- right now we have one reservation object per ipset, which makes it simple, also for the cleanup. To reduce number of objects we could test to store the reservations per subnet, or AllocationRange? as this CRD is "internal" we could change this after initial landing.

hjensas commented:
> @hjensas I think I addressed all/most of your comments. Thanks for those! A lot of the complexity in reading is now removed. I think its small and clean. For this initial version I will do tomorrow
>
>     * add all relevant information for the consumer to the ipset status, so that all network details could fetched from there instead of query the netconfig for it, you'd agree on that?
> 

Yes, that sounds ok to me.

hjensas commented on apis/network/v1beta1/netconfig_types.go:
```diff
@@ -0,0 +1,178 @@
+	// +kubebuilder:validation:Optional
+	// ExcludeAddresses a set of IPs that should be excluded from range
+	// TODO: add validation to webhook
+	ExcludeAddresses []string `json:"excludeAddresses,omitempty"`
```
I know I asked for exclude's in allocation pool earlier, but when I look at it again now I think it would it be better to move the ExcludeAddresses to the subnet struct? Since we use it both for in allocation range and outside allocation range in fixedIPs case, feels like the UX is better if this is per subnet - not per-allocation range.

If we set it at Subnet, and place in the AssignIpDetails struct? Code in FixedIPs would become simpler as well since no need to iterate AllocationRanges.

stuggi replied:
ack sounds good

stuggi replied:
done

hjensas reviewed (commented):
+1 I think this is close to something we can merge and iterate on.
One new suggestion related to ExcludeAddresses in-line.

rabi commented on apis/network/v1beta1/reservation_types.go:
```diff
@@ -0,0 +1,79 @@
+// ReservationSpec defines the desired state of Reservation
+type ReservationSpec struct {
+	// Hostname
+	Hostname *HostNameStr `json:"hostname"`
```
We probably don't need this as we've IPSetRef?

stuggi replied:
yes, could do that. I thought its easier to identify for any troubleshooting which host this is mapped to

stuggi replied:
done, removed hostname

rabi commented on apis/network/v1beta1/ipset_types.go:
```diff
@@ -0,0 +1,94 @@
+// IPSetSpec defines the desired state of IPSet
+type IPSetSpec struct {
+	// Hostname
+	Hostname *HostNameStr `json:"hostname"`
```
Do we need this? Won't the requester be the owner of this?

stuggi replied:
right now we use the hostname to validate for an already reserved IP. you suggest we drop the hostname from ipam at all and use the IPSetRef for this?

rabi replied:
Yeah I meant that, reservation is for a IPSet, whether it's created by dataplane role/node or a controlplane service.

stuggi replied:
yes, I think that makes sense. the IPSetRef is a clear reference to the IPSet and not having the hostname is more generic and the requestor handles it.

stuggi replied:
done, removed hostname

stuggi commented on apis/network/v1beta1/netconfig_types.go:
```diff
@@ -0,0 +1,178 @@
+// NetReservation - Reservations per subnet
+type NetReservation struct {
+	Subnet NetNameStr `json:"subnet,omitempty" optional:"true"`
+	Count  int        `json:"count,omitempty" optional:"true"`
```
its not used right now. initial plan was to count how many IPs from the AllocationRange are in use. let me remove the status from NetConfig for now and we add it when its actually used

stuggi commented on apis/network/v1beta1/netconfig_types.go:
```diff
@@ -0,0 +1,178 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +kubebuilder:validation:Pattern="^[a-zA-Z0-9][a-zA-Z0-9\\-_]*[a-zA-Z0-9]$"
```
sure we can do the validation in the webhook of the network naming, but I am not sure what you mean with metadata/name. this is for the netname, not the cr name.

rabi replied:
Ignore that. For a moment I assumed that this CRD is one per network, hence cr name = netname.

stuggi commented on controllers/network/ipset_controller.go:
```diff
@@ -0,0 +1,394 @@
+			Reservelist: reservations,
+		}
+		if ipsetNet.FixedIP != nil {
+			ipDetails.FixedIP = net.ParseIP(string(*ipsetNet.FixedIP))
```
missed that, will add it

rabi commented:
I've not looked at the controller code (assumed only CRDs based on the commit message). I can  see ipset_controller.go in the last patch. Probably more patches to be added?

stuggi commented:
> I've not looked at the controller code (assumed only CRDs based on the commit message). I can see ipset_controller.go in the last patch. Probably more patches to be added?

right now the ipset_controller is the only required controller. netconfig and reservation will only have webhooks, at least  for now

rabi commented:
> > I've not looked at the controller code (assumed only CRDs based on the commit message). I can see ipset_controller.go in the last patch. Probably more patches to be added?
> 
> right now the ipset_controller is the only required controller. netconfig and reservation will only have webhooks, at least for now

Hmm.. Based on our earlier discussion  in [design doc](https://docs.google.com/document/d/1TLK_fph10lRh1Bs1t-LwRmcz8OK7qEvDsLsoBhTYmHo/edit#) I expected NetConfig controller to do the actual reservation. Any reason we did not go with that?

`NetConfig controller watches IPSets
Validates if the networks requested matches the configured ones
Optional requested IPs are valid, available and no duplicates
Generates IPs per requested network and creates reservation object for the IPSet
Adds IPSet as additional owner to auto-reconcile on changes`


This is somewhat similar to what I had initially proposed other than additional IPReservation CRDs, no(?)

stuggi commented:
> > > I've not looked at the controller code (assumed only CRDs based on the commit message). I can see ipset_controller.go in the last patch. Probably more patches to be added?
> > 
> > 
> > right now the ipset_controller is the only required controller. netconfig and reservation will only have webhooks, at least for now
> 
> Hmm.. Based on our earlier discussion in [design doc](https://docs.google.com/document/d/1TLK_fph10lRh1Bs1t-LwRmcz8OK7qEvDsLsoBhTYmHo/edit#) I expected NetConfig controller to do the actual reservation. Any reason we did not go with that?
> 
> `NetConfig controller watches IPSets Validates if the networks requested matches the configured ones Optional requested IPs are valid, available and no duplicates Generates IPs per requested network and creates reservation object for the IPSet Adds IPSet as additional owner to auto-reconcile on changes`
> 
> This is somewhat similar to what I had initially proposed other than additional IPReservation CRD, no(?)

right, sorry for the confusion. first I did the implementation as in the doc. the ipset controller had to watch all IPSets, as we know could be a lot. due to the discussion on scaling/performance, I switch back to be really similar to what you proposed as I now think it probably is more efficient with less reconciliation.

abays commented on controllers/network/ipset_controller.go:
```diff
@@ -0,0 +1,446 @@
+		return nil
+	})
+
+	// 'UpdateFunc' used to judge if a event about the object is what we want.
+	// If that is true, the event will be processed by the reconciler.
+	p := predicate.Funcs{
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			return e.ObjectOld.GetResourceVersion() != e.ObjectNew.GetResourceVersion()
+		},
+	}
```
If `CreateFunc` and `DeleteFunc` are excluded as they are here, does that mean they implicitly return true?

stuggi replied:
oh you are correct, I misunderstood it. thinking of  it again, we could probably remove the predicate func

stuggi replied:
done

abays commented on controllers/network/ipset_controller.go:
```diff
@@ -0,0 +1,446 @@
+	r.Log.Info("Reconciling Service delete")
+
+	// get reservation
+	res, err := r.getReservation(ctx, instance)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
```
I think it's possible to get a `nil` `res` here given L320-323 (but `nil` `err` as well, so it won't return on L208).  Maybe we should check for that before calling `controllerutil.RemoveFinalizer()`?

stuggi replied:
ack, will address that. probably change getReservation and handle NotFound in the upper level

stuggi replied:
done

abays commented on controllers/network/ipset_controller.go:
```diff
@@ -0,0 +1,446 @@
+		return ctrl.Result{}, nil
+	}
+
+	instance.Status.Reservation = []networkv1.IPAddress{}
```
Should we check for `instance.Status.Reservation == nil` here before setting?  Just curious.  I know we do this elsewhere as a pattern for maps, but not sure for arrays/slices.

stuggi replied:
we always reconstruct the status.Reservation, so I think its good to always start with an empty reservation

abays replied:
I'm just wondering about a case where we hit an error in the reconcile loop over and over at a certain spot in code before the reservations would be calculated.  i.e. we set the `Reservations` to an empty array here but hit an error before populating them.  The `defer` status update would still fire and update the CR to have empty reservations.  I'm just curious whether this could confuse some other resource that was watching the `IPSet` CR.  But maybe it's not a problem.

stuggi replied:
ack got your point. In such case the ipset will have an overall `Read` condition which is `False`. The requestor should not use it until the condition is `True`. I think that will prevent this situation?

abays replied:
Okay, that seems fair enough to me

stuggi commented on apis/network/v1beta1/conditions.go:
```diff
@@ -0,0 +1,51 @@
+	ReservationErrorMessage = "Reservation error occured %s"
+
+	// ReservationMissMatchErrorMessage
+	ReservationMissMatchErrorMessage = "Reservation reservations (%d) do not match requested (%d)"
```
done

stuggi commented:
added what was on TODO:

```
* add all relevant information for the consumer to the ipset status,
  so that all network details could fetched from there instead of query
  the netconfig for it
```

stuggi commented:
/test infra-operator-build-deploy

stuggi commented:
build-deploy job hit bundle size in https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openstack-k8s-operators_infra-operator/40/pull-ci-openstack-k8s-operators-infra-operator-main-infra-operator-build-deploy/1661632885234339840/artifacts/infra-operator-build-deploy/openstack-k8s-operators-gather/artifacts/must-gather/quay-io-openshift-release-dev-ocp-v4-0-art-dev-sha256-6eac618c70508e7c135084b46d87f2a293ed5ea9d4ca616280f0663afb1a3e77/namespaces/openshift-operator-lifecycle-manager/pods/catalog-operator-595d4787cd-nq2rd/catalog-operator/catalog-operator/logs/current.log


`2023-05-25T07:50:19.853024151Z E0525 07:50:19.852955       1 queueinformer_operator.go:290] sync {"update" "openstack/install-fbzt5"} failed: failed to update installplan bundle lookups: etcdserver: request is too large`

rabi reviewed (commented):
I'll give it a spin and do a round of tests tomorrow with the test-cases I had prepared for my initial PR. If we want to merge it in the meantime I'm fine though. We can iterate on it later.

stuggi commented on controllers/network/ipset_controller.go:
```diff
@@ -0,0 +1,429 @@
+		}
+
+		// For each ipsets instance create a reconcile request
+		for _, i := range ipsets.Items {
```
as from above example, with this version the status of the ipset has all the details the requestor need. if there are changed, like routes get added to the netconfig, the ipsets reconciles to reflect that info.

stuggi commented on controllers/network/ipset_controller.go:
```diff
@@ -0,0 +1,429 @@
+	netcfgs := &networkv1.NetConfigList{}
+	err := r.List(ctx, netcfgs, opts)
+	if err != nil {
+		instance.Status.Conditions.Set(condition.FalseCondition(
```
yes could also use this. I can change it

stuggi replied:
done

abays commented:
https://github.com/openstack-k8s-operators/openstack-operator/pull/324 merged.  Retesting.

/test infra-operator-build-deploy

abays commented:
> @stuggi: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:
> Test name 	Commit 	Details 	Required 	Rerun command
> ci/prow/infra-operator-build-deploy 	[04013c8](https://github.com/openstack-k8s-operators/infra-operator/commit/04013c8bc0e9f115da19b57e468f8e4bffd75589) 	[link](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openstack-k8s-operators_infra-operator/40/pull-ci-openstack-k8s-operators-infra-operator-main-infra-operator-build-deploy/1661737145666113536) 	false 	`/test infra-operator-build-deploy`
> 
> [Full PR test history](https://prow.ci.openshift.org/pr-history?org=openstack-k8s-operators&repo=infra-operator&pr=40). [Your PR dashboard](https://prow.ci.openshift.org/pr?query=is:pr+state:open+author:stuggi).

Logs make it seem like there's a problem trying to install the new separated RabbitMQ bundle:

```
      message: bundle contents have not yet been persisted to installplan status
      reason: BundleNotUnpacked
      status: "True"
      type: BundleLookupNotPersisted
    - lastTransitionTime: "2023-05-25T14:49:25Z"
      message: unpack job not completed
      reason: JobIncomplete
      status: "True"
      type: BundleLookupPending
    identifier: cluster-operator.v0.0.0
    path: quay.io/openstack-k8s-operators/rabbitmq-cluster-operator-bundle@sha256:9ff91ad3c9ef1797b232fce2f9adf6ede5c3421163bff5b8a2a462c6a2b3a68b
    properties: '{"properties":[{"type":"olm.gvk","value":{"group":"rabbitmq.com","kind":"RabbitmqCluster","version":"v1beta1"}},{"type":"olm.package","value":{"packageName":"cluster-operator","version":"0.0.0"}}]}'
```

stuggi commented:
docs PR https://github.com/openstack-k8s-operators/docs/pull/33

stuggi commented:
/test infra-operator-build-deploy

stuggi commented:
@dprince I think we basically hit the sized limit again with adding the new CRDs here (which are not big). 

`2023-05-26T14:11:10.996467430Z E0526 14:11:10.996422       1 queueinformer_operator.go:290] sync {"update" "openstack/install-nghgn"} failed: failed to update installplan bundle lookups: etcdserver: request is too large`

https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openstack-k8s-operators_infra-operator/40/pull-ci-openstack-k8s-operators-infra-operator-main-infra-operator-build-deploy/1662091326222503936/artifacts/infra-operator-build-deploy/openstack-k8s-operators-gather/artifacts/must-gather/quay-io-openshift-release-dev-ocp-v4-0-art-dev-sha256-6eac618c70508e7c135084b46d87f2a293ed5ea9d4ca616280f0663afb1a3e77/namespaces/openshift-operator-lifecycle-manager/pods/catalog-operator-595d4787cd-6wb9q/catalog-operator/catalog-operator/logs/current.log

The same seem to happen in this heat PR to the openstack-operator [1]
[1] https://github.com/openstack-k8s-operators/openstack-operator/pull/321

abays commented:
> @dprince I think we basically hit the sized limit again with adding the new CRDs here (which are not big).
> 
> `2023-05-26T14:11:10.996467430Z E0526 14:11:10.996422 1 queueinformer_operator.go:290] sync {"update" "openstack/install-nghgn"} failed: failed to update installplan bundle lookups: etcdserver: request is too large`
> 
> https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openstack-k8s-operators_infra-operator/40/pull-ci-openstack-k8s-operators-infra-operator-main-infra-operator-build-deploy/1662091326222503936/artifacts/infra-operator-build-deploy/openstack-k8s-operators-gather/artifacts/must-gather/quay-io-openshift-release-dev-ocp-v4-0-art-dev-sha256-6eac618c70508e7c135084b46d87f2a293ed5ea9d4ca616280f0663afb1a3e77/namespaces/openshift-operator-lifecycle-manager/pods/catalog-operator-595d4787cd-6wb9q/catalog-operator/catalog-operator/logs/current.log
> 
> The same seem to happen in this heat PR to the openstack-operator [1] [1] [openstack-k8s-operators/openstack-operator#321](https://github.com/openstack-k8s-operators/openstack-operator/pull/321)

I was able to reproduce this locally, even with all the latest OpenStack operator bits where we had removed Rabbit from the bundle to conserve space.

abays commented:
/test infra-operator-build-deploy
